tests/kernel/pipe/pipe_api/src/test_pipe_bench.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2016 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <kernel_structs.h>
+#include <ksched.h>
+#include <wait_q.h>
+
+#if CONFIG_X86
+#define TIMING_INFO_PRE_READ()
+#define TIMING_INFO_OS_GET_TIME() (_tsc_read())
+#elif CONFIG_ARM
+#define TIMING_INFO_PRE_READ()
+#define TIMING_INFO_OS_GET_TIME()      (k_cycle_get_32())
+#endif
+
+#define STACK_SIZE      1024
+#define PIPE_LEN 16
+K_PIPE_DEFINE(kpipe4, PIPE_LEN, 4);
+K_THREAD_STACK_DEFINE(tstack4, STACK_SIZE);
+K_THREAD_STACK_DEFINE(tstack5, STACK_SIZE);
+K_THREAD_STACK_DEFINE(tstack6, STACK_SIZE);
+K_THREAD_STACK_DEFINE(tstack7, STACK_SIZE);
+__kernel struct k_thread tdata4;
+__kernel struct k_thread tdata5;
+__kernel struct k_thread tdata6;
+__kernel struct k_thread tdata7;
+
+
+static void thread_handler(void *p1, void *p2, void *p3)
+{
+	while (1) {
+		k_yield();
+	}
+}
+
+void test_bench(void)
+{
+	struct k_thread *thread;
+
+	k_tid_t tid1 = k_thread_create(&tdata4, tstack4, STACK_SIZE,
+					thread_handler, NULL, NULL, NULL,
+					K_PRIO_PREEMPT(1), 0, 0);
+
+	k_tid_t tid2 = k_thread_create(&tdata5, tstack5, STACK_SIZE,
+					thread_handler, NULL, NULL, NULL,
+					K_PRIO_PREEMPT(2), 0, 0);
+
+	k_tid_t tid3 = k_thread_create(&tdata6, tstack6, STACK_SIZE,
+					thread_handler, NULL, NULL, NULL,
+					K_PRIO_PREEMPT(3), 0, 0);
+
+	k_tid_t tid4 = k_thread_create(&tdata7, tstack7, STACK_SIZE,
+					thread_handler, &kpipe4, NULL, NULL,
+					K_PRIO_PREEMPT(4), 0, 0);
+
+	k_sleep(10);
+
+	TIMING_INFO_PRE_READ();
+	u32_t pre_pend_time1 = TIMING_INFO_OS_GET_TIME();
+
+	_pend_thread((struct k_thread *) tid1,
+			&kpipe4.wait_q.writers, K_NO_WAIT);
```

Comment:
FWIW: this isn't legal.   If you try it with assertions enabled, it should fail as tid1 is neither the current thread or a DUMMY thread.  Basically, you can only pend yourself (because in SMP you don't know that thread isn't already running).  This usage looks like it won't blow up, but this isn't the right way to do this.

Basically: this is measuring the overhead of an operation that never happens in practice.